### PR TITLE
webapp: add custom metrics around GitHub HTTP API usage

### PR DESCRIPTION
**Pull request by jgehrcke**
This is for milestone 5 in https://github.com/conbench/conbench/issues/493#issuecomment-1413770175.

### CODE REVIEW

File: conbench/__init__.py [hunk 1]
```diff
@@ -79,45 +74,12 @@ def create_application(config):
...
-
-    if not path:
-        return
-
-    try:
-        log.info("os.path.isdir('%s'): %s", path, os.path.isdir(path))
-    except OSError as exc:
-        log.info("os.path.isdir('%s') failed: %s", path, exc)
-
-
```
Comment (jgehrcke) [inline conbench/__init__.py:120]: Moved this logic unaltered to `conbech/metrics.py`.

File: requirements-webapp.txt [hunk 2]
```diff
@@ -16,6 +16,7 @@ marshmallow
 numpy
 oauthlib
 openapi-spec-validator
+prometheus-client
```
Comment (jgehrcke) [inline requirements-webapp.txt:19]: This was an implicit dependency of `prometheus-flask-exporter`. Better to make it explicit.

Not so important. At some point I think we will have all dependencies explicitly locked down (pinned) anyway.

File: conbench/__init__.py [hunk 3]
```diff
@@ -33,16 +33,11 @@
 conbench.logger.setup(level_stderr="DEBUG", level_file=None, level_sqlalchemy="WARNING")
 log = logging.getLogger(__name__)
 
-# This is going to be an application-global singleton (in the webapp, not the
-# CLI).
-metrics = None
```
Comment (jgehrcke) [inline conbench/__init__.py:38]: I removed this again, it was premature thinking.

**Comment by jonkeane** [comment]
I know that there's a little bit of an ordering conundrum (need this in prod to show it with prod data), but do you have any screenshots of what these would look like from a local setup? Mostly I'm curious about how these custom metrics align with the flask ones that we're not defining manually

**Comment by jgehrcke** [comment]
> but do you have any screenshots of what these would look like from a local setup? Mostly I'm curious about how these custom metrics align with the flask ones that we're not defining manually

Thank you for asking! :) Once these metrics are defined and exposed and get scraped they first all only get written into Prometheus' time series database. Then Grafana _can_ be used to query those specific time series and to build dashboard panels based on those metrics.

The querying/dashboarding/panel-building/alerting work is of course required to get some value from these metrics.

But it's also an independent chunk of work.

I am currently in fact working on some basic panel-building on my machine with a dummy/local/mock setup. 

Here is a screenshot showing Grafana's panel builder (served by Grafana running on k8s/minikube) and me having manually entered a Prometheus query referring to one of the new metrics (pink arrow):

Screenshot from 2023-02-09 17-50-09

Will submit the outcome of this dashboarding work as another PR, showing a screenshot of the Conbench dashboard then (having some new panels).

File: conbench/__init__.py (hunk 3 above)
Comment (austin3dickey) [reply conbench/__init__.py:38]: Nice, I'm scared of globals. :)

File: conbench/metrics.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def decorate_flask_app_with_metrics(app) -> None:
+    """
+    Add flask-prometheus-exporter magic to `app`.
+
+    This mutates `app` in-place.
+    """
+    # Use `GunicornPrometheusMetrics` when spawning a separate HTTP server for
+    # the metrics scrape endpoing. Note that this sets the global singleton.
```
Comment (austin3dickey) [inline conbench/metrics.py:88]: nit: there's no longer a global singleton

**Review by austin3dickey — APPROVED** [review]
Lookin good! Excited to see the dashboards

File: conbench/metrics.py (hunk 4 above)
Comment (jgehrcke) [reply conbench/metrics.py:88]: Thank you for paying attention!

Will update this in the next patch.

**Comment by jgehrcke** [comment]
>  I'm curious about how these custom metrics align with the flask ones that we're not defining manually

@jonkeane quick follow-up: https://github.com/conbench/conbench/pull/722 has a screenshot that sheds a bit of light onto that. 

At some point in the near-ish future we should have demo time and walk through these topics a little more slowly to make the most of this.

**Comment by jonkeane** [comment]
Thanks, those look good, and does get at some of what I was wondering (mostly: what _are_ the magic auto captured flask things and how much coverage does that have on it's own — they're both important, but I hope that the magic auto capture stuff is pretty significant for the obvious things that our API is doing) and a lot of that would be great for:

> At some point in the near-ish future we should have demo time and walk through these topics a little more slowly to make the most of this.

Yes please! So we can all learn a bit more about this stuff!

File: conbench/metrics.py (hunk 4 above)
Comment (jgehrcke) [reply conbench/metrics.py:88]: done here: https://github.com/conbench/conbench/pull/725

**Pull request merged**